src/datahike/query.cljc

```diff
@@ -1033,191 +1055,537 @@ in those cases.
+                               (step# dst-inner#
+                                      ~@(map (fn [i sym]
+                                               (if (nil? i)
+                                                 sym
+                                                 `(nth ~substitution-value-vector ~i)))
+                                             pmask
+                                             pattern-symbols)
+                                      ~pred-expr))
+                             dst#
+                             ~subst-filt-pairs)))))]
+
+         ;; Generate different code depending on whether or not there is a
+         ;; `filt-extractor`, meaning that the resulting datoms have to be
+         ;; filtered.
+         `(if (nil? ~filt-extractor)
+            ~(branch-expr datom-pred)
+            ~(branch-expr `(extend-predicate1 ~datom-pred ~filt-extractor ~filt))))))))
+
+(defn instantiate-substitution-xform [substitution-pattern-element-inds
+                                      filt-extractor
+                                      subst-filt-pairs]
+
+  ;; Returns a transducer based on the indices in `substitution-pattern-element-inds`
+  (substitution-expansion substitution-pattern-element-inds
+                          filt-extractor
+                          subst-filt-pairs))
+
+;; The performance improvement of using this macro has been measured,
+;; see comment in single-substition-xform.
+(defmacro make-vec-lookup-ref-replacer [range-length]
+  (let [inds (gensym)
+        replacer (gensym)
+        tuple (gensym)]
+    `(fn tree-fn# [~replacer ~inds]
+       ~(dt/range-subset-tree
+         range-length inds
+         (fn replacer-fn# [pinds _mask]
+           `(fn [~tuple]
+              (try
+                ~(mapv (fn [index i] `(~replacer ~index (nth ~tuple ~i)))
+                       pinds
+                       (range))
+                (catch Exception e# nil))))))))
+
+(def vec-lookup-ref-replacer (make-vec-lookup-ref-replacer 5))
+
+;; The performance improvement of using this macro has been measured,
+;; see comment in single-substition-xform.
+(defmacro basic-index-selector [max-length]
+  (let [inds (gensym)
+
+        obj (gensym)]
+    `(fn [~inds]
+       (case (count ~inds)
+         ~@(mapcat (fn [length]
+                     (let [index-symbols (vec (repeatedly length gensym))]
+                       [length `(let [~index-symbols ~inds]
+                                  (fn [~obj]
+                                    ~(mapv
+                                      (fn [sym] `(nth ~obj ~sym))
+                                      index-symbols)))]))
+                   (range (inc max-length)))))))
+
+(def make-basic-index-selector (basic-index-selector 5))
+
+(defn single-substitution-xform
+  "Returns a transducer that substitutes the symbols for a single relation."
+  [search-context relation-index subst-map filt-map]
```

Comment:
I should think about better names instead of `subst-map` and `filt-map`. This code is a bit complicated so better names would help for the sake of readability.

Reply:
Right, that helps in general, but it is also ok if they are explained where they are introduced, or renamed to shorthands in internal scopes.